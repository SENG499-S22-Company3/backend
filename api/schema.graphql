type Query {
  """
  Get the current user
  """
  me: User
  """
  Find a user by their id
  """
  findUserById(id: ID!): User
  """
  Get Teaching Preference Survey
  """
  survey: TeachingPreferenceSurvey!
}

type Mutation {
  """
  Login into a user account using email and password
  """
  login(username: String!, password: String!): AuthPayload!
  """
  Logout the currently logged in user
  """
  logout: AuthPayload!
  """
  Register a new user account
  """
  createUser(username: String!): RegisterUserMutationResult!
  """
  Updates a user given the user id.
  """
  updateUser(input: UpdateUserInput!): UpdateUserMutationResult
  """
  Change the password of the currently logged in user
  """
  changeUserPassword(
    input: ChangeUserPasswordInput!
  ): ChangeUserPasswordMutationResult!
  """
  Reset a users password.
  """
  resetPassword(id: ID!): ResetPasswordMutationResult!
  """
  Teaching preferences
  """
  createTeachingPreference(
    input: CreateTeachingPreferenceInput!
  ): CreateTeachingPreferenceMutationResult!
}

input CreateTeachingPreferenceInput {
  peng: Boolean!
  userId: ID!
  courses: [CoursePreference!]!
}

type TeachingPreferenceSurvey {
  courses: [CourseID!]!
}

type CoursePreference implements CourseID {
  course: CourseID!
  preference: Int!
}

type CourseID {
  subject: String!
  number: String!
  term: Term!
}

enum Term {
  Fall
  Spring
  Summer
}

input ChangeUserPasswordInput {
  oldPassword: String!
  newPassword: String!
}

type ChangeUserPasswordMutationResult {
  success: Boolean!
  message: String
}

type ResetPasswordMutationResult {
  success: Boolean!
  message: String
  password: String
}

input UpdateUserInput {
  id: ID!
}

type UpdateUserMutationResult {
  user: User
  errors: [Error!]
}

type CreateUserMutationResult {
  success: Boolean!
  message: String
  username: String
  password: String
}

"""
Returned when a user logs in our logs out.
"""
type AuthPayload {
  """
  Whether auth operation was successful or not
  """
  success: Boolean!
  """
  Optional error message
  """
  message: String
}

"""
User
"""
type User {
  id: ID!
  username: String!
  firstName: String!
  lastName: String!
  role: Role!
  """
  Determine if the user is marked active
  """
  active: Boolean!
}

"""
User role
"""
enum Role {
  """
  Administrator role (department staff etc.)
  """
  ADMIN
  """
  User role (professor, student etc.)
  """
  USER
}