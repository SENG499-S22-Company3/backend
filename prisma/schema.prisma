// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider      = "prisma-client-js"
  binaryTargets = "native"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id            Int                  @id @default(autoincrement())
  username      String               @unique
  password      String
  displayName   String?
  active        Boolean
  hasPeng       Boolean
  role          Role                 @default(USER)
  courseSection CourseSection[]
  preference    TeachingPreference[]
}

model Schedule {
  id            Int             @id @default(autoincrement())
  year          Int
  createdOn     DateTime        @default(now())
  courseSection CourseSection[]
}

model CourseSection {
  id            Int           @id @default(autoincrement())
  hoursPerWeek  Float
  sectionNumber String
  capacity      Int
  startDate     DateTime
  endDate       DateTime
  meetingTime   MeetingTime[]
  user          User?         @relation(fields: [userId], references: [id])
  userId        Int?
  course        Course        @relation(fields: [courseId], references: [id])
  courseId      Int
  schedule      Schedule?     @relation(fields: [scheduleId], references: [id])
  scheduleId    Int?
}

model Course {
  id               Int                @id @default(autoincrement())
  subject          String
  courseCode       String
  title            String
  streamSequence   String
  sengRatio        Float              @default(0.75)
  capacity         Int                @default(0)
  term             Term
  courseSection    CourseSection[]
  coursePreference CoursePreference[]

  @@unique([subject, courseCode, term])
}

model TeachingPreference {
  id                 Int                @id @default(autoincrement())
  coursePreference   CoursePreference[]
  nonTeachingTerm1   Term?
  nonTeachingTerm2   Term?
  hasRelief          Boolean
  reliefReason       String?
  studyLeave         Boolean
  studyLeaveReason   String?
  topicsOrGradCourse Boolean
  topicDescription   String?
  fallTermCourses    Int                @default(2)
  springTermCourses  Int                @default(2)
  summerTermCourses  Int                @default(2)
  User               User?              @relation(fields: [userId], references: [id])
  userId             Int?               @unique

}

model CoursePreference {
  id                Int                 @id @default(autoincrement())
  preference        Int
  course            Course              @relation(fields: [courseId], references: [id])
  courseId          Int
  teachPreference   TeachingPreference? @relation(fields: [teachPreferenceId], references: [id])
  teachPreferenceId Int?

  @@unique([teachPreferenceId, preference, courseId])
}

model MeetingTime {
  id              Int            @id @default(autoincrement())
  startTime       DateTime
  endTime         DateTime
  days            Day[]
  courseSection   CourseSection? @relation(fields: [courseSectionId], references: [id])
  courseSectionId Int?

  @@unique([startTime, endTime, days, courseSectionId])
}

enum Role {
  USER
  ADMIN
}

enum Day {
  MONDAY
  TUESDAY
  WEDNESDAY
  THURSDAY
  FRIDAY
  SATURDAY
  SUNDAY
}

enum Term {
  FALL
  SPRING
  SUMMER
}
