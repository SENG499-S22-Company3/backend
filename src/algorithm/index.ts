import axios, { AxiosInstance } from 'axios';

import { SchedulePostRequest, Schedule } from '../client/algorithm1';
import { CourseObject } from '../client/algorithm2';
import { Company } from '../schema';

// these are the default ports for each algo and company
export const algoUrl = {
  algorithm1: {
    company3: process.env.ALGO1_COMPANY_3_URL || 'http://localhost:8080',
    company4: process.env.ALGO1_COMPANY_4_URL || 'http://localhost:8000',
  },
  algorithm2: {
    company3: process.env.ALGO2_COMPANY_3_URL || 'http://localhost:5000',
    company4: process.env.ALGO2_COMPANY_4_URL || 'http://localhost:5000',
  },
};
// after importing the generated api

const createSchedule =
  (instance: AxiosInstance) => (input: SchedulePostRequest) => {
    return instance.post<Schedule>('/schedule', input);
  };

const createPredict = (instance: AxiosInstance) => (input: CourseObject[]) => {
  return instance.post<CourseObject[]>('/predict_class_size', input);
};

const checkSchedule =
  (instance: AxiosInstance) => (input: SchedulePostRequest) => {
    return instance.post<String>('/checkSchedule', input);
  };

const algo1c3Cs = checkSchedule(
  axios.create({
    baseURL: algoUrl.algorithm1.company3,
  })
);

const algo1c3 = createSchedule(
  axios.create({
    baseURL: algoUrl.algorithm1.company3,
  })
);

const algo1c4 = createSchedule(
  axios.create({
    baseURL: algoUrl.algorithm1.company4,
  })
);

const algo2c3 = createPredict(
  axios.create({
    baseURL: algoUrl.algorithm2.company3,
  })
);

const algo2c4 = createPredict(
  axios.create({
    baseURL: algoUrl.algorithm2.company4,
  })
);

export const request = (company: Company) => {
  if (company === 'COMPANY3') {
    return {
      algo1: algo1c3,
      algo2: algo2c3,
      algo1Cs: algo1c3Cs,
    };
  }
  return {
    algo1: algo1c4,
    algo2: algo2c4,
  };
};

console.info('Setting up Algorithm APIs...');
console.info('Algorithm 1 (Company 3) ->', algoUrl.algorithm1.company3);
console.info('Algorithm 1 (Company 4) ->', algoUrl.algorithm1.company4);
console.info('Algorithm 2 (Company 3) ->', algoUrl.algorithm2.company3);
console.info('Algorithm 2 (Company 4) ->', algoUrl.algorithm2.company4);
