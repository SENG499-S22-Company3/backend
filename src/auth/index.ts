import { Prisma, User } from '@prisma/client';
import { prisma, lookupUser } from '../prisma';
import bcrypt from 'bcrypt';
import type {
  AuthPayload,
  ChangeUserPasswordInput,
  Response,
  CreateUserMutationResult,
} from '../schema/graphql';
import type { Context } from '../context';
export { login, changePassword, createNewUser };

// User with username testuser and password testpassword
// username: 'testuser',
// password: '$2b$10$ogZBif.TabQ/LoAk8LjlG./hNq3tsWBE9OAzbc.dY/hQdYMIPhBly',

const failedAuth = {
  message: 'Incorrect username or password',
  success: false,
  token: '', // Won't be needed anymore
};

async function login(
  ctx: Context,
  username: string,
  password: string
): Promise<AuthPayload> {
  const user = await lookupUser(username);

  if (user === null) return failedAuth;

  const valid = await bcrypt.compare(password, user.password);

  if (!valid) {
    return {
      message: 'Incorrect username or password',
      success: false,
      token: '', // Won't be needed anymore
    };
  } else {
    ctx.session.user = user;
    return {
      message: 'Success',
      success: true,
      token: '',
    };
  }
}

async function changePassword(
  user: User,
  { currentPassword, newPassword }: ChangeUserPasswordInput
): Promise<Response> {
  const valid = await bcrypt.compare(currentPassword, user.password);
  if (!valid) {
    return {
      message: 'Incorrect previous password',
      success: false,
    };
  } else {
    console.log('UPDATING PASSWORD TO', newPassword);
    const pwsalt = bcrypt.genSaltSync(10);
    const pwhash = bcrypt.hashSync(newPassword, pwsalt);
    await prisma.user.update({
      where: {
        username: user.username,
      },
      data: {
        password: pwhash,
      },
    });
    return {
      message: 'Password Changed Successfully',
      success: true,
    };
  }
}

async function createNewUser(
  username: string
): Promise<CreateUserMutationResult> {
  try {
    await prisma.user.create({
      data: {
        username: username,
        password:
          '$2b$10$ogZBif.TabQ/LoAk8LjlG./hNq3tsWBE9OAzbc.dY/hQdYMIPhBly',
        active: false,
        hasPeng: false,
      },
    });
  } catch (e: unknown) {
    // P2002 is the code for unique constraint violation
    if (
      e instanceof Prisma.PrismaClientKnownRequestError &&
      e.code === 'P2002'
    ) {
      return {
        message: `Failed to create user, username '${username}' already exists.`,
        success: false,
      };
    } else {
      throw e;
    }
  }

  return {
    message:
      "User created successfully with password 'testpassword'. Please login and change it.",
    success: true,
    username: username,
    password: 'testpassword',
  };
}
