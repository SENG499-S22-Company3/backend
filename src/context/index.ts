import { Request } from 'express';
import { Session, SessionData } from 'express-session';

export interface Context {
  // TODO: keep here until we have login implemented. a partial given it could be undefined
  session: Session & Partial<SessionData>;
  // TODO: add context (aka. prisma, auth, etc.)
  logout: () => Promise<void>;
}

export async function createContext({
  req,
}: {
  req: Request;
}): Promise<Context> {
  // TODO: add any request specific context (aka. prisma, auth, etc.)
  // TODO: check for session with cookie
  const { session } = req;
  return {
    session,
    logout: async () => new Promise((resolve) => session.destroy(resolve)),
  };
}
