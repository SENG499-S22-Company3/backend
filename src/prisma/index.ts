import { PrismaClient } from '@prisma/client';
export { prisma };

const prisma = new PrismaClient();
