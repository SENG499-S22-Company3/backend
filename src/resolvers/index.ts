import { AxiosError, AxiosResponse } from 'axios';
import {
  changePassword,
  createNewUser,
  generateSchedule,
  login,
  resetPassword,
} from '../auth';
import { Schedule } from '../client/algorithm1';
import { CourseObject } from '../client/algorithm2';
import { Context } from '../context';
import { getAllCourses } from '../prisma/course';
import { findUserById } from '../prisma/user';
import { Resolvers, Term } from '../schema';
import * as utils from '../utils';
import {
  createSchedule,
  getAll,
  getCourses,
  getMe,
  getSchedule,
  getUserByID,
  prepareCourseCapacities,
  prepareScheduleWithCapacities,
  updateUserSurvey,
} from './resolverUtils';

const noLogin = {
  success: false,
  message: 'Not logged in',
  token: '',
};

const alreadyLoggedIn = {
  success: false,
  message: 'Already logged in',
  token: '',
};

const noPerms = {
  message: 'Insufficient permisions',
  success: false,
};

const apiErrorHandler = (id: string, e: unknown, data: any = {}) => {
  console.error(id, e);
  if (e instanceof AxiosError) {
    const error = {
      message: `${id}: API call error: Bad Response:${e.message}:${e.response?.data}`,
      status: e.response?.status,
      ...data,
    };
    return {
      message: JSON.stringify(error),
      success: false,
    };
  } else {
    const error = {
      message: `${id}: Backend ran into an internal server error:\n${e}`,
    };
    return {
      message: JSON.stringify(error),
      success: false,
    };
  }
};

export const resolvers: Resolvers<Context> = {
  Query: {
    me: async (_, _params, ctx) => {
      if (!ctx.user) return null;
      return await getMe(ctx.user);
    },
    findUserById: async (_, params, ctx) => {
      if (!ctx.user || !params.id) return null;
      return await getUserByID(+params.id);
    },
    survey: async (_, __, ctx) => {
      if (!ctx.user) return { courses: [] };

      const courses = (await getAllCourses()).map((c) => ({
        ...c,
        code: c.courseCode,
        term:
          c.term === 'FALL'
            ? Term.Fall
            : c.term === 'SPRING'
            ? Term.Spring
            : Term.Summer,
      }));

      return {
        courses,
      };
    },
    courses: async (_, params, ctx) => {
      if (!ctx.user || !params.term) return null;
      return await getCourses(params.term);
    },
    schedule: async (_, params, ctx) => {
      if (!ctx.user) return null;
      return getSchedule(params.year || new Date().getFullYear());
    },
    allUsers: async (_, _params, ctx) => {
      if (!ctx.user || !utils.isAdmin(ctx.user)) return null;
      return await getAll();
    },
  },
  Mutation: {
    login: async (_, params, ctx) => {
      if (ctx.user) return alreadyLoggedIn;
      return await login(ctx, params.username, params.password);
    },
    logout: async () => {
      return {
        success: true,
        message: 'logged out',
        token: '',
      };
    },
    changeUserPassword: async (_, _params, ctx) => {
      if (!ctx.user) return noLogin;
      return changePassword(ctx.user, _params.input);
    },
    resetPassword: async (_, _params, ctx) => {
      if (!ctx.user) return noLogin;
      else if (!utils.isAdmin(ctx.user)) return noPerms;
      return await resetPassword(_params.id);
    },
    createUser: async (_, { username }, ctx) => {
      if (!ctx.user) return noLogin;
      else if (!utils.isAdmin(ctx.user)) return noPerms;
      return await createNewUser(username);
    },
    createTeachingPreference: async (_, { input }, ctx) => {
      if (!ctx.user) return noLogin;

      if (ctx.user.preference.length !== 0) {
        return {
          token: '',
          success: false,
          message:
            'Teaching preferences survey has already been submitted for this user',
        };
      }

      await updateUserSurvey(ctx.user.id, input);

      // refresh the context user when they update their preferences
      const refreshedUser = await findUserById(ctx.user.id);
      // should never be null but just in case
      if (refreshedUser !== null) {
        ctx.user = refreshedUser;
      }

      return {
        token: '',
        success: true,
        message: 'Teaching preferences updated.',
      };
    },
    generateSchedule: async (_, { input }, ctx) => {
      if (!ctx.user) return noLogin;
      else if (!utils.isAdmin(ctx.user)) return noPerms; // Only Admin can generate schedule

      const users = await getAll();

      let capacityDataResponse: AxiosResponse<CourseObject[], any> | null;
      const algo2Payload = prepareCourseCapacities(input);
      try {
        if (algo2Payload === null) {
          return {
            success: false,
            message: 'No courses selected',
          };
        }

        capacityDataResponse = await ctx
          .algorithm(input.algorithm2)
          .algo2(algo2Payload);
      } catch (e) {
        return apiErrorHandler(`ALGORITHM2_${input.algorithm2}`, e);
      }

      if (!capacityDataResponse) {
        return {
          success: false,
          message: 'Error: No Courses Input',
        };
      }

      console.log('ALG 2 RESPONSE DATA');
      console.log(capacityDataResponse.data);
      console.log('END ALG 2 RESPONSE DATA');

      let scheduleResponse: AxiosResponse<Schedule, any> | null;
      const algo1Payload = await prepareScheduleWithCapacities(
        input,
        users,
        capacityDataResponse.data
      );
      try {
        scheduleResponse = await ctx
          .algorithm(input.algorithm1)
          .algo1(algo1Payload);
      } catch (e) {
        return apiErrorHandler(`ALGORITHM1_${input.algorithm1}`, e, {
          algorithm2: {
            request: algo2Payload,
            response: capacityDataResponse.data,
          },
          algorithm1: {
            request: algo1Payload,
          },
        });
      }

      if (!scheduleResponse?.data) {
        return {
          success: false,
          message: 'Error: No response from algorithm 1',
        };
      }

      console.log('ALG 1 RESPONSE DATA');
      console.log(scheduleResponse.data);
      console.log('END ALG 1 RESPONSE DATA');

      try {
        await createSchedule(input.year, scheduleResponse.data);
      } catch (e) {
        console.log(`Failed to generate schedule: '${e}'`);
        return {
          success: false,
          message: 'backend failed to insert generated schedule into database',
        };
      }

      return generateSchedule(input);
    },
  },
};
