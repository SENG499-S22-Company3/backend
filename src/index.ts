import { ApolloServerPluginDrainHttpServer } from 'apollo-server-core';
import { ApolloServer } from 'apollo-server-express';
import express, { Express } from 'express';
import fs from 'fs';
import http from 'http';
import path from 'path';
import { algoUrl } from './algorithm';
import { createContext } from './context';
import { resolvers } from './resolvers';
import { Resolvers } from './schema';
const isProduction = process.env.NODE_ENV === 'production';
const port = process.env.PORT || 4000;
const schemaPath = path.join(__dirname, 'schema/schema.graphql');

async function readSchema(schemaPath: string) {
  return await fs.promises.readFile(schemaPath, 'utf8');
}

// https://www.apollographql.com/docs/apollo-server/integrations/middleware/#apollo-server-express
async function start(app: Express, typeDefs: any, resolvers: Resolvers) {
  const httpServer = http.createServer(app);

  const server = new ApolloServer({
    typeDefs,
    resolvers,
    // csrfPrevention: true,
    // eslint-disable-next-line new-cap
    plugins: [ApolloServerPluginDrainHttpServer({ httpServer })],
    // TODO: add context (aka. prisma, auth, etc.)
    context: createContext,
  });

  await server.start();
  server.applyMiddleware({
    app,
    cors: {
      origin: ['https://studio.apollographql.com'],
      credentials: true,
    },
  });

  await new Promise<void>((resolve) => httpServer.listen({ port }, resolve));

  console.log(
    `ðŸš€ Server ready at http://localhost:${port}${server.graphqlPath}`
  );
}

const app = express();
isProduction && app.set('trust proxy', 1);

app.get('/healthcheck', (_req, res) => {
  res.send('OK');
});

app.get('/runtime', (_req, res) => {
  res.json(algoUrl);
});

async function main() {
  // read schema
  const schema = await readSchema(schemaPath);

  // attach Apollo Server start the web server
  start(app, schema, resolvers);
}

main();
