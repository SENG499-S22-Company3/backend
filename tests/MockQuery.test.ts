import { ApolloServer, ExpressContext } from 'apollo-server-express';
import { mocks } from './mockDatasource';
import { Term } from '../src/schema';
import fs from 'fs';
import {
  expectUser,
  expectSurvey,
  expectCourses,
  expectSchedule,
} from './mockexpect';
import {
  meQuery,
  finduseridquery,
  surveyQuery,
  coursesquery,
  schedulequery,
} from './typedefs';

const typeDefs = fs.readFileSync('./src/schema/schema.graphql', 'utf8');
let testServer: ApolloServer<ExpressContext>;

beforeAll(() => {
  testServer = new ApolloServer({
    typeDefs,
    mocks,
  });
});

describe('QUERY MOCK TESTS', () => {
  it('Mocking me', async () => {
    const result = await testServer.executeOperation({
      query: meQuery,
    });

    expectUser(result.data?.me);
    expect(result.data).toMatchSnapshot();
  });

  it('Mocking survey', async () => {
    const result = await testServer.executeOperation({
      query: surveyQuery,
    });

    expectSurvey(result.data?.survey.courses[0]);
    expect(result.data).toMatchSnapshot();
  });

  it('Mocking findUserByID', async () => {
    const result = await testServer.executeOperation({
      query: finduseridquery,
    });

    expectUser(result.data?.findUserById);
  });

  it('Mocking Courses', async () => {
    const result = await testServer.executeOperation({
      query: coursesquery,
      variables: { term: Term.Fall },
    });

    expectCourses(result.data?.courses[0]);
    expect(result.data).toMatchSnapshot();
  });

  it('Mocking Schedule', async () => {
    const result = await testServer.executeOperation({
      query: schedulequery,
      variables: { year: 2021, term: Term.Fall },
    });

    expectSchedule(result.data?.schedule);
    expect(result.data).toMatchSnapshot();
  });
});
