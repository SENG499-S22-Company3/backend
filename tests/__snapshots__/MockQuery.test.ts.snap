// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QUERY MOCK TESTS Mocking Courses 1`] = `
Object {
  "courses": Array [
    Object {
      "CourseID": Object {
        "code": "159.92.65.89",
        "subject": "Feil LLC",
        "term": "FALL",
      },
      "capacity": 20029,
      "endDate": "s<}OBCgCDy",
      "hoursPerWeek": 43086,
      "meetingTimes": Array [
        Object {
          "day": "FRIDAY",
          "endTime": "Tp63?3y)7^",
          "startTime": "6YrK^sVZ[j",
        },
      ],
      "professors": Array [
        Object {
          "active": true,
          "id": 97681,
          "password": "Investment Account",
          "preferences": Array [
            Object {
              "id": Object {
                "code": "221.82.180.4",
                "subject": "Mayer, Dach and Welch",
                "term": "FALL",
              },
              "preference": 30725,
            },
          ],
          "role": "USER",
          "username": "Graham",
        },
      ],
      "startDate": "uu?Hu&Htq&",
    },
    Object {
      "CourseID": Object {
        "code": "174.54.38.60",
        "subject": "Lowe - Lesch",
        "term": "FALL",
      },
      "capacity": 55907,
      "endDate": "(Kr.g_jaXV",
      "hoursPerWeek": 32957,
      "meetingTimes": Array [
        Object {
          "day": "FRIDAY",
          "endTime": "G9]yHv\\"A6y",
          "startTime": "/M,Z]_)uJJ",
        },
      ],
      "professors": Array [
        Object {
          "active": true,
          "id": 67266,
          "password": "Personal Loan Account",
          "preferences": Array [
            Object {
              "id": Object {
                "code": "221.82.180.4",
                "subject": "Mayer, Dach and Welch",
                "term": "FALL",
              },
              "preference": 30725,
            },
          ],
          "role": "USER",
          "username": "Georgette",
        },
      ],
      "startDate": "};{IV/B<,w",
    },
  ],
}
`;

exports[`QUERY MOCK TESTS Mocking Schedule 1`] = `
Object {
  "schedule": Object {
    "courses": Array [
      Object {
        "CourseID": Object {
          "code": "150.130.71.90",
          "subject": "Ullrich Group",
          "term": "FALL",
        },
        "capacity": 48741,
        "endDate": "#xdz?Ff^vW",
        "hoursPerWeek": 68523,
        "meetingTimes": Array [
          Object {
            "day": "FRIDAY",
            "endTime": "1_c%evIJc.",
            "startTime": "rvr''iVSTK",
          },
        ],
        "professors": Array [
          Object {
            "active": true,
            "id": 57382,
            "password": "Credit Card Account",
            "preferences": Array [
              Object {
                "id": Object {
                  "code": "221.82.180.4",
                  "subject": "Mayer, Dach and Welch",
                  "term": "FALL",
                },
                "preference": 30725,
              },
            ],
            "role": "USER",
            "username": "Boyd",
          },
        ],
        "startDate": "]KotXv\\"Ccr",
      },
    ],
    "createdAt": "IbI^qQLvKt",
    "id": "Bahamas",
    "year": 3,
  },
}
`;

exports[`QUERY MOCK TESTS Mocking me 1`] = `
Object {
  "me": Object {
    "active": true,
    "id": 74186,
    "password": "Savings Account",
    "preferences": Array [
      Object {
        "id": Object {
          "code": "221.82.180.4",
          "subject": "Mayer, Dach and Welch",
          "term": "FALL",
        },
        "preference": 30725,
      },
    ],
    "role": "USER",
    "username": "Joanne",
  },
}
`;

exports[`QUERY MOCK TESTS Mocking survey 1`] = `
Object {
  "survey": Object {
    "courses": Array [
      Object {
        "code": "93.213.43.160",
        "subject": "Metz, Prosacco and Rutherford",
        "term": "FALL",
      },
    ],
  },
}
`;
